allennlp/models/semantic_parsing/text2sql_parser.py

```diff
@@ -0,0 +1,399 @@
+        return predicted_tensor.equal(targets_trimmed)
+
+    @staticmethod
+    def is_nonterminal(token: str):
+        if token[0] == '"' and token[-1] == '"':
+            return False
+        return True
+
+    @overrides
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        """
+        We track four metrics here:
+
+            1. exact_match, which is the percentage of the time that our best output action sequence
+            matches the SQL query exactly.
+
+            2. denotation_acc, which is the percentage of examples where we get the correct
+            denotation.  This is the typical "accuracy" metric, and it is what you should usually
+            report in an experimental result.  You need to be careful, though, that you're
+            computing this on the full data, and not just the subset that can be parsed. (make sure
+            you pass "keep_if_unparseable=True" to the dataset reader, which we do for validation data,
+            but not training data).
+
+            3. valid_sql_query, which is the percentage of time that decoding actually produces a
+            valid SQL query.  We might not produce a valid SQL query if the decoder gets
+            into a repetitive loop, or we're trying to produce a super long SQL query and run
+            out of time steps, or something.
+
+            4. action_similarity, which is how similar the action sequence predicted is to the actual
+            action sequence. This is basically a soft measure of exact_match.
+        """
+        return {
+                'exact_match': self._exact_match.get_metric(reset),
+                'denotation_acc': self._denotation_accuracy.get_metric(reset),
+                'valid_sql_query': self._valid_sql_query.get_metric(reset),
+                'action_similarity': self._action_similarity.get_metric(reset)
+                }
+
+    def _create_grammar_state(self, possible_actions: List[ProductionRule]) -> GrammarStatelet:
+        """
+        This method creates the GrammarStatelet object that's used for decoding.  Part of creating
+        that is creating the `valid_actions` dictionary, which contains embedded representations of
+        all of the valid actions.  So, we create that here as well.
+
+        The inputs to this method are for a `single instance in the batch`; none of the tensors we
+        create here are batched.  We grab the global action ids from the input
+        ``ProductionRules``, and we use those to embed the valid actions for every
+        non-terminal type.  We use the input ``linking_scores`` for non-global actions.
+
+        Parameters
+        ----------
+        possible_actions : ``List[ProductionRule]``
+            From the input to ``forward`` for a single batch instance.
+        """
+        device = util.get_device_of(self._action_embedder.weight)
+        # TODO(Mark): This type is pure \(- . ^)/
```

Comment:
Yeah...  Maybe we should introduce a class for this...?  That might make it easier.